Automation listing UI updates 1/3 - MAILPOET-4532

johnolek opened the pull request:
This PR implements a number of UI tweaks to the automations listing page.

johnolek committed:
Remove actions dropdown

MAILPOET-4532

johnolek committed:
Add MailPoet top bar to automation listing page

MAILPOET-4532

johnolek committed:
Add button to create new automation and extract wrapper

MAILPOET-4532

johnolek committed:
Add tab filters and implement pagination

[MAILPOET-4532]

johnolek committed:
Redirect to trash if attempting to edit trashed workflow

[MAILPOET-4532]

johnolek commented on mailpoet/assets/js/src/automation/automation.tsx:
```diff
@@ -79,24 +102,27 @@ function DeleteSchemaButton(): JSX.Element {
-        </CreateWorkflowFromTemplateButton>
-        <RecreateSchemaButton />
-        <DeleteSchemaButton />
+    <BrowserRouter>
```
I don't know if we want/need to wrap the entire App with the router

JanJakes replied:
My guess is the router only needs to sit above the highest place in the hierarchy where we're routing. That said, it's probably fine to have it somewhere very close to the root of the app or directly at the root, so I think this is good.

johnolek commented on mailpoet/assets/js/src/automation/listing/index.tsx:
```diff
@@ -1,18 +1,90 @@
 
+type FilterTab = {
+  name: string;
+  title?: JSX.Element;
```
Only nullable because of how I build the tabs later, probably unnecessary.

johnolek commented on mailpoet/assets/js/src/automation/listing/index.tsx:
```diff
@@ -21,56 +93,64 @@ export function AutomationListing({ workflows, loading }: Props): JSX.Element {
-            )}
+    <TabPanel
+      className="mailpoet-filter-tab-panel"
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore - the Tab type actually expects a string for titles but won't render HTML,
+      // making it very difficult to style the count badges. It seems to be compatible wtih JSX
+      // elements, however.
+      tabs={tabs}
```
This might be one of the biggest issues. I'm not sure how we would get the styled badge counts if we have to use a normal string for the tab titles

JanJakes replied:
This [looks safe](https://github.com/WordPress/gutenberg/blob/39172d1aa22adfa0b337de6bb53801276f868065/packages/components/src/tab-panel/index.js#L87), at least for now. Maybe [the DefinitelyTyped types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/fb66211010980641f01bcfe8ffbc0cc3d71989ca/types/wordpress__components/tab-panel/index.d.ts#L12) are even wrong.

There are also some [other fields](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/fb66211010980641f01bcfe8ffbc0cc3d71989ca/types/wordpress__components/tab-panel/index.d.ts#L21) but I guess those help only in the tab body and not in the title?

JanJakes replied:
This might be solvable also via CSS, using some pseudoelement like `::after` and passing in value from some data attribute, maybe. But it's weird and probably not very accessible.

johnolek replied:
I would prefer to avoid a CSS solution like that if possible, but I also don't want this to cause any issues, or really require the `@ts-ignore` at all.

> There are also some other fields but I guess those help only in the tab body and not in the title?

This matches my understanding. You can attach arbitrary key/value pairs but they're only intended to be used in the callback that renders the tab body.

I've refactored to an approach that doesn't require using JSX for the `title` (even though it's probably totally fine), but I doubt that it's doing things the "react" way. I would appreciate your guidance on how to make it better if you have any ideas!

JanJakes replied:
> but I doubt that it's doing things the "react" way

Ahh, interesting idea, but yeah, this is really dangerous in React! You're creating elements under the React tree that are not known to React (not rendered within React's VDOM). Then any diffing will simply not count with those — they may be accidentally lost or cause some different problems due to undefined behavior. In this case, the previous `@ts-ignore` solution actually looks better and safer to me.

Again, we may try to create an issue [in the upstream repo to together with a PR changing the type](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/tab-panel/types.ts) (a funny thing I just noticed is they converted it to TS a few days ago).

But we should probably note these things in our checklist doc and do them post-MVP.

johnolek replied:
Very good points! I've converted back to the JSX for now.

JanJakes commented on mailpoet/assets/js/src/automation/automation.tsx:
```diff
@@ -11,20 +14,39 @@ import { WorkflowListingNotices } from './listing/workflow-listing-notices';
+        <Onboarding />
+      ) : (
+        <AutomationListing workflows={workflows} loading={loading} />
+      );
```
Nothing too important, but if you extract this to a component (can even be local to this file), you can use early returns instead of the non-const variable.

johnolek replied:
Great idea, thanks for the suggestion!

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_listing.scss:
```diff
@@ -2,6 +2,22 @@
 
+.mailpoet-filter-tab-panel {
+  background-color: #fff;
+  border-radius: calc(2px - 1px);
```
Isn't this `1px`? 😄 Or just a leftover from some experiments?

johnolek replied:
Haha, sorry about this! I was copying CSS rules from my browser to mimic the visual effects of the table. 

At first I thought there might just be a simple CSS class I could re-use, but it looks like the class name for these styles is computed, and I'm not sure where it's coming from or how I could re-use this manually. I'm sure there's a better way to handle this.

Here are the styles I'm talking about from my inspector:

<img width="567" alt="image" src="https://user-images.githubusercontent.com/8656640/188198700-f58046a2-e092-4a60-bea7-786cc748b529.png">

JanJakes commented on mailpoet/assets/js/src/automation/listing/index.tsx:
```diff
@@ -1,17 +1,90 @@
+      }
+    });
+    history.replace({ search: pageSearch.toString() });
+  };
```
Just to be sure, I'd wrap this in `useCallback()` so the function doesn't get redefined with each render. Not really important though, I guess.

johnolek replied:
I made this change, and also added `useMemo` for the `pageSearch` variable, but I'm not sure how much of a difference it will make. Won't it get redefined every time anyway since it will have `pageSearch` in its dependency array? Maybe it's useful to have for later, when we add other features and wouldn't want to redefine this callback if they make changes that don't affect the query string for the page?

JanJakes replied:
The weird thing about React is that it kind of changes the semantics of the language, or at least how people understand it. A function suddenly must be expected to be called (rendered) many times, even for a single "render" from the user's perspective. That's not only non-performant but also very unintuitive, and it often causes developers to wrap a lot of stuff in `useMemo` and `useCallback` calls.

One may think that we should use these hooks only for some computationally expensive logic, but that's not the case either — when you want a stable reference to some object (for instance, to avoid re-rendering children components via changing props), then you need memoization as well.

This is why frameworks like [Solid.js](https://www.solidjs.com/) or [Svelte](https://svelte.dev/) are so much more performant, and, in many ways, better.

While in many cases, you may not need the hooks, I tend to use them rather more than less as a "good practice". As for the `pageSearch` — sure, in itself it would be a very quick computation, and if the identity of the object doesn't cascade many further renders, then it's probably OK without `useMemo`. But since the result is not a scalar value and it's identity would change with every render, it may be nice to keep the memo to prevent any possible issues.

Note that for a single "view" for the user, the component function is sometimes called multiple times — each `setState`, `useEffect`, or prop changes coming from parents can cause another re-render that the user won't notice (unless you start randomly losing focus in an editor, etc.). So if you type `console.log('render');` to some of our component bodies, you may see it logged multiple times. This stands true until advanced batching and some form of auto-memoization are in React itself.

johnolek replied:
Thank you very much for this detailed explanation!

JanJakes commented on mailpoet/assets/js/src/automation/listing/index.tsx:
```diff
@@ -1,17 +1,90 @@
+      groupedWorkflows[workflow.status] = [];
+    }
+    groupedWorkflows[workflow.status].push(workflow);
+  });
```
Here, I'd probably use `useMemo` to avoid recomputing it on every render.

There's also a declarative way to write this, although not sure if more readable:
```ts
const groupedWorkflows = workflows.reduce(
  (grouped, workflow) => ({
    ...grouped,
    [workflow.status]: [...(grouped[workflow.status] ?? []), workflow],
  }),
  {},
);
```

johnolek replied:
Good call on `useMemo`!

This is mostly a personal failing on my part, but I've never liked using `reduce`. It was hard for me to understand when I first started coding and I've never warmed up to it.

In this case I'm curious, would the performance be the same using the spread syntax compared to `push`? I did a little bit of testing and I think spread option is slower, but it probably wouldn't make any kind of noticeable difference without there being tons and tons of workflows to filter.

I tested with 15k workflows and your code took ~700ms to sort compared to ~1ms for using `.push`. The much, much slower part was actually the query to retrieve all the workflows in the first place because there's no `workflow_id` index on the `workflow_versions` table, so I'm going to add one.

I'm happy to change to something more declarative but I'd like to avoid the performance hit if possible.

JanJakes replied:
Good catch 😎  Reduce with spread inside surely is slower in cases like these because it has a complexity of `O(n^2)` (for each `n`, you need to iterate and reconstruct the already-existing array), compared to the possible `O(n)`, not to mention the extra temporary allocations. For a reasonably small amount of items, though, this is probably not measurable (as we'd never return thousands of workflows from the API). I would also guess the language optimizer could resolve these cases to something simpler, but as you've measured, it doesn't seem so.

In React state management, this approach is extremely common — most things are written declaratively to make sure (and visible) the proper objects and sub-objects have new instances and never modify the original ones. Often this is also solved using [Immer](https://github.com/immerjs/immer) and other libs.

Readability is probably partly a habit and personal preference. I guess that `reduce` in particular is a bit difficult, but otherwise, I got used to the declarative style quite a bit and find it readable.

In this case, use which option you prefer. Ultimately, both work if you don't modify the original object, and yes, yours may be more readable here.

JanJakes commented on mailpoet/assets/js/src/automation/listing/index.tsx:
```diff
@@ -1,17 +1,90 @@
+    }
+
+    tabs.push(tab);
+  });
```
Here, it may be nicer to extract the status to a component. I'd add `useMemo` again and maybe make it declarative again (with `tabLabels.map`).

Generally, there is a lot of logic directly in the component render body, but that's normal when prototyping. Later, it's better use add `useMemo` and `useCallback` or even better — extract some of it to a custom hook.

johnolek replied:
I refactored this using `useMemo` and a `map`, but I'm not sure I completely follow what it would entail to convert it to its own component or use a custom hook. Could you please elaborate a little bit about what that might look like so I can further improve this?

JanJakes replied:
For instance, the workflow grouping can be extracted simply as follows:

```ts
const useGroupedWorkflows = (
  workflows: Workflow[],
): Record<string, Workflow[]> => {
  return useMemo(() => {
    const grouped = {};
    workflows.forEach((workflow) => {
      if (!grouped[workflow.status]) {
        grouped[workflow.status] = [];
      }
      grouped[workflow.status].push(workflow);
    });
    return grouped;
  }, [workflows]);
};
```

This is a custom hook, it's as simple as that. Thanks to `useMemo`, it returns a stable identity for the same `workflows` parameter.

---

As for tabs and table headers — they don't have any dependencies from the state or component body, so they could be simple module-level arrays (indeed ever allocated only once):

```ts
const tabs = [
  {
    name: 'all',
    title: 'All',
    className: 'mailpoet-tab-all',
  },
  {
    name: WorkflowStatus.ACTIVE,
    title: 'Active',
    className: 'mailpoet-tab-active',
  },
  {
    name: WorkflowStatus.INACTIVE,
    title: 'Inactive',
    className: 'mailpoet-tab-inactive',
  },
  {
    name: WorkflowStatus.DRAFT,
    title: 'Draft',
    className: 'mailpoet-tab-draft',
  },
  {
    name: WorkflowStatus.TRASH,
    title: 'Trash',
    className: 'mailpoet-tab-trash',
  },
] as const;
```

I even simplified it by inlining the `className`, seems more redable to me now. (Another nice thing is we can add the `as const` and TS will now understand that `label: 'Trash'` belongs to `className: 'mailpoet-trash-tab'` instad of inferring `label: string` and `className` string. But we don't need this in the code below so not important.)

Similarly for the table headers...

johnolek replied:
> This is a custom hook, it's as simple as that

Is the benefit here that we would be able to extract the callback outside of this component entirely? If so where would be the best place for it to live? Or is there a different advantage that I'm not seeing?

johnolek replied:
Thank you again for your detailed explanation! I wasn't familiar with `as const` before.

JanJakes replied:
> Is the benefit here that we would be able to extract the callback outside of this component entirely? If so where would be the best place for it to live? Or is there a different advantage that I'm not seeing?

I think in this case it's only about code organization, just giving that part of code a function name (hook name) and making it slightly more readable. I'd probably just put it in the same file. For custom hooks used in multiple places, we'd need to extract them to some specific files, I guess, but this one is used only locally.

React docs encourage the usage of custom hooks, it can partly separate business logic from rendering, but if it's not meant to be reused in multiple places, then there are not many more advantages than that. When naming conventions are met (the `use` prefix), linters also understand that it is a hook and enforce the [hook rules](https://reactjs.org/docs/hooks-rules.html).

So ultimately, it's just a tiny detail, nothing we'd have to do now.

johnolek replied:
> So ultimately, it's just a tiny detail, nothing we'd have to do now

Still great to know, thanks for the explanation!

JanJakes commented on mailpoet/assets/js/src/automation/listing/index.tsx:
```diff
@@ -1,17 +1,90 @@
+    Object.keys(keysAndValues).forEach((key) => {
+      pageSearch.set(key, keysAndValues[key]);
+      if (['per_page', 'status'].includes(key)) {
+        pageSearch.set('paged', '1');
```
One thing I noticed is WP doesn't seem to show `paged=1` for posts. They only start from `2` but for `1` they normalize it out.

johnolek replied:
Great catch! I'll delete the parameter instead of setting it to 1 in this case.

JanJakes reviewed (commented):
Looks good! I think it's almost ready.

johnolek committed:
Refactor content to be a local component

MAILPOET-4532

johnolek committed:
Delete "paged" search param instead of setting to 1

MAILPOET-4532

johnolek committed:
Use useMemo to avoid recalculating groups every render

MAILPOET-4532

johnolek committed:
Add workflow_id index to versions table

MAILPOET-4532

johnolek committed:
Refactor tab calculation

MAILPOET-4532

johnolek committed:
Add counts dynamically to avoid type mismatch

MAILPOET-4532

johnolek committed:
Use useCallback for query string updater

MAILPOET-4532

johnolek committed:
Restore original button padding

The .has-icon class for some reason redefines the padding to be `6x`
instead of `6px 12px`. This was making the button look almost cut off on
 the right side when we used an icon.

MAILPOET-4532

johnolek committed:
"Fix" some styling issues

This is a somewhat hacky approach to getting tabs to render as desired.
Because the table content appears inside the tab container it's
difficult to have matching box shadows because they overlap. There were
also rounded corners at the top of the border underneath the tabs.

MAILPOET-4532

johnolek committed:
Memoize more things

MAILPOET-4532

johnolek commented on mailpoet/assets/js/src/automation/listing/index.tsx:
```diff
@@ -1,76 +1,180 @@
+          pageSearch.delete('paged');
+        }
+      });
+      history.push({ search: pageSearch.toString() });
```
This is the one thing I'm still really struggling with. I'm not sure how to update the current tab in the TabPanel's state when a user hits back/forward. Do you have any ideas for how to handle this so users can actually navigate forwards/backwards through the tabs?

JanJakes replied:
I checked the [TabPanel component](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/tab-panel), and weirdly enough, they provide only `initialTabName` but no `activeTabName` to make its state controlled. Maybe we should create an issue for that on their repo.

In the meantime, you can use a hack — nothing nice but very simple — add `key={pageSearch.get('status')}` to the `TabPanel` component. This will force-unmount and mount the component and together with the `initialTabName={pageSearch.get('status') || 'all'}` it will make it work.

JanJakes replied:
On a different note, I'm not sure why do you check also for the status part in this case (`['per_page', 'status'].includes(key)`), those params seem unrelated, aren't they?

The loop seems to modify the original `pageSearch` object (which in React might not be the best to do) and in general seems a bit hard to read. I think what you mean is "take existing params + new params + delete paged when `1`" which can be written as:

```ts
const updateUrlSearchString = useCallback(
  (search: Record<string, string>) => {
    const params = new URLSearchParams({
      ...Object.fromEntries(pageSearch.entries()),
      ...search,
    });
    if (params.get('paged') === '1') {
      params.delete('paged');
    }
    history.push({ search: params.toString() });
  },
  [pageSearch, history],
);
```

johnolek replied:
Thank you for the `key` suggestion! Not ideal, but better than having something non-functional.

> On a different note, I'm not sure why do you check also for the status part in this case (['per_page', 'status'].includes(key)), those params seem unrelated, aren't they?

The idea is to reset to the first page when a user clicks on a different tab or changes the number of results per page. If you're viewing page 2 of "All" and then click on "Inactive" you wouldn't necessarily want to immediately view page 2 of the "Inactive" tab. And if there aren't enough "Inactive" workflows for there to be a second page, the pagination controls won't even appear and you can end up seeing something like this:

<img width="1355" alt="image" src="https://user-images.githubusercontent.com/8656640/188716990-2b1daf3a-b2f6-46d8-84e9-2715ca76300e.png">

It's similar but different for `per_page`. Let's say you have 240 workflows and you're viewing page 5/10 with 25 workflows per page. If you switch to 100 per page there are now only 3 total pages, so leaving `paged=5` would result in an empty table. 

We could try to dynamically calculate a logical new page number. For example we could figure out what the first workflow was on the page you were viewing previously and then switch to whichever page had that workflow with the new `per_page` value. If you were previously looking at workflows 101-125 then we could figure out we should show you 101-200 with the new `per_page`. That just seemed overly complicated and may not even be what users expect.

JanJakes replied:
Ahh, thanks for the explanation, I totally missed that!

johnolek commented:
Thank you so much for the detailed review, @JanJakes! I think I've fixed some things and I've probably introduced other problems. Please let me know if you have any other ideas, especially for how to get history working.

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_listing.scss:
```diff
@@ -1,3 +1,34 @@
+    /* Prevent border radius beneath tabs */
+    border-radius: 0 0 1px 1px;
+    /* Prevent overlapping box shadow since listing is rendered inside tabs */
+    clip-path: inset(-1px 0 0 0);
```
I wonder whether it's a good idea to clip out the `box-shadow` rather than resetting the box shadow itself. Or maybe I'm not understanding what exactly this does. Can the listing have all borders and box shadows reset and can we apply them only on the outer tabs, maybe?

Anyway, the result looks good 👍 So maybe no need to fix anything here.

johnolek replied:
The intention here is to clip out all of the box shadow except for 1px of shadow on the top.

It would be possible to use a border instead of a box shadow, but then I'm not sure if there's an elegant way to get the bottom border of the active tab to cover the border.

Using border:
<img width="424" alt="image" src="https://user-images.githubusercontent.com/8656640/188710977-1eec236e-1968-4638-847c-e89c23939caf.png">

Using box shadow with clipping:
<img width="493" alt="image" src="https://user-images.githubusercontent.com/8656640/188710906-daa247f1-71fb-4562-af08-12f07a4d8113.png">

I think it actually makes a pretty big difference zoomed out (this is just me switching between the two versions a few times):

https://user-images.githubusercontent.com/8656640/188711373-b6d5e405-60be-492d-b963-d428847ebe60.mov

I think I'll just leave this as-is for now and hopefully we can come up with a cleaner solution in the future when we refactor this into a proper component.

JanJakes replied:
Ahh, OK! I didn't see this is more complex than just borders.

johnolek committed:
Enable browser forward/back between tabs

MAILPOET-4532

johnolek committed:
Rewrite URL search updater

Create a new object instead of modifying the original, and remove the
`paged` parameter when viewing the first page of results.

MAILPOET-4532

johnolek committed:
Revert to safer way of generating filter count badges

The previous method was unsafe due to its dependence on creating
elements outside what React knew about

MAILPOET-4532

johnolek committed:
Only update the status query string when necessary

This prevents pushing duplicate entries to the browser history. For
example, if you are on the "All" tab, then click the "Inactive" tab 3
times, you shouldn't have to hit the back button 3 times to get back to
"All," it should only require a single back operation.

MAILPOET-4532

johnolek committed:
Add type to memoized grouped workflows

MAILPOET-4532

johnolek committed:
Extract consts where possible and simplify

MAILPOET-4532

JanJakes commented:
LGTM! I did a rebase to resolve a conflict. Merging.
